src/components/shared/Swiper.tsx

```diff
@@ -0,0 +1,329 @@
+  pagingEnabled?: boolean;
+  showsHorizontalScrollIndicator?: boolean;
+  showsVerticalScrollIndicator?: boolean;
+  bounces?: boolean;
+  scrollsToTop?: boolean;
+  removeClippedSubviews?: boolean;
+  automaticallyAdjustContentInsets?: boolean;
+  index?: number;
+  children: Element[];
+};
+
+interface SwiperState {
+  isScrolling: boolean;
+  offset: number;
+  total: number;
+  index: number;
+  width: number;
+  height: number;
+}
+
+export default class Swiper extends Component<SwiperProps, SwiperState> {
+  private scrollView?: ScrollView | null;
+
+  // Props for ScrollView component
+  static defaultProps = {
+    horizontal: true,
+    pagingEnabled: true,
+    showsHorizontalScrollIndicator: false,
+    showsVerticalScrollIndicator: false,
+    bounces: false,
+    scrollsToTop: false,
+    removeClippedSubviews: true,
+    automaticallyAdjustContentInsets: false,
+    index: 0
+  };
+
+  state = this.initState(this.props);
+
+  shouldComponentUpdate(
+    nextProps: SwiperProps,
+    nextState: SwiperState
+  ): boolean {
+    return (
+      this.state.isScrolling !== nextState.isScrolling ||
+      this.state.offset !== nextState.offset
+    );
+  }
+
+  /**
+   * Initialize the state
+   */
+  initState(props: SwiperProps): SwiperState {
+    // Get the total number of slides passed as children
+    const total = props.children ? props.children.length || 1 : 0,
+      // Current index
+      index =
+        total > 1 ? Math.min(props.index ? props.index : 0, total - 1) : 0,
+      // Current offset
+      offset = width * index;
+
+    const state = {
+      total,
+      index,
+      offset,
+      width,
+      height,
+      isScrolling: false
+    };
+
+    return state;
+  }
+
+  /**
+   * Scroll begin handler
+   * @param {object} e native event
+   */
+  onScrollBegin: ScrollViewProps["onScrollBeginDrag"] = (
+    event?: NativeSyntheticEvent<NativeScrollEvent>
+  ) => {
+    if (!event) {
+      // TODO: what does this case even mean? read react-native docs
+      return;
+    }
+    // Update internal isScrolling state
+    this.setState({ isScrolling: true });
+  };
+
+  /**
+   * Scroll end handler
+   */
+  onScrollEnd: ScrollViewProps["onMomentumScrollEnd"] = event => {
+    if (!event) {
+      // TODO: what does this case even mean? read react-native docs
+      return;
+    }
+    // Update internal isScrolling state
+    this.state.isScrolling = false;
+
+    // Update index
+    this.updateIndex(
+      event.nativeEvent.contentOffset
+        ? event.nativeEvent.contentOffset.x
+        : this.state.offset
+      // TODO: When scrolled with .scrollTo() on Android there is no contentOffset
+      // event.nativeEvent.position * this.state.width
+    );
+  };
+
+  /*
+     * Drag end handler
+     */
+  onScrollEndDrag: ScrollViewProps["onScrollEndDrag"] = event => {
+    if (!event) {
+      // TODO: what does this case even mean? read react-native docs
+      return;
+    }
+    const {
```

Comment:
Similar here about just doing these as separate const calls, it just looks odd to me lol